### PR TITLE
#1727 CD Pipeline Tag Versioning

### PR DESCRIPTION
**Pull request by ldraney**
# Description

Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. List any dependencies that are required for this change.

issue #1727 

The goal of this PR is to automate tagging as a part of our release process. 

The values from the initial PR merge to master will be saved in the workflow; and once it has been merged to the release branch, those labels with be the basis for the tag version that is created on the release branch. 

The latest SHA on the release branch will be used for the new tag, which will be semver bumped according to the following logic:
Major version bump: breaking change
Patch version bump: hotfix, security, bug fix
Minor version bump: every other label

And appropriate checks and GitHub Step Summaries are included to ensure visibility to QA of what's happening behind the scenes. 

[Octokit](https://octokit.github.io/rest.js/v20), in context of[ github-script](https://github.com/actions/github-script), was key to accomplishing this work. 

### Choosing a trigger
I first thought that a trigger related to pull_request would be used here;
However, since that trigger goes off whenever a PR is closed (merged or not), I learned that the [trigger to use on a pull request merge is `push`](https://github.com/orgs/community/discussions/26724), and then to [grab PR information from push events](https://github.com/orgs/community/discussions/27071). 

Thus the logical flow was:
1. Grab the commit SHA from the [push trigger payload](https://docs.github.com/en/webhooks/webhook-events-and-payloads#push)
2. [Use that commit SHA](https://octokit.github.io/rest.js/v20#repos-list-pull-requests-associated-with-commit) to find the [PR number and labels](https://docs.github.com/en/rest/commits/commits?apiVersion=2022-11-28#list-pull-requests-associated-with-a-commit).  Only one PR should be associated with the commit.
3. Use the labels from the PR object for semver versioning

### A note on the Refactor
@kalbfled  Provided some valuable feedback for refactoring that was applied.  One of these was the noted duplicate coded for each step summary.  This inspired me to start a actionUtils.js file, which I intend to be a central location for avoiding duplicate code and modularizing functionality for future situations like this.  The relevant function here is appendSummary.

## Type of change

Please check the relevant option(s).

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update
- [ ] Documentation changes only

## How Has This Been Tested?

How has this been tested? (e.g. Unit tests, Tested locally, Tested as a GitHub action, Depoyed to dev, etc)  
Please provide relevant information and / or links to demonstrate the functionality or fix. (e.g. screenshots, link to deployment, regression test results, etc)

[Latest POC test](https://github.com/department-of-veterans-affairs/notification-api/actions/runs/8974925166)

### tests after refactor
[PR Label Semver Summary](https://github.com/department-of-veterans-affairs/notification-api/actions/runs/9009041271)
[Pre Tag Release Summary](https://github.com/department-of-veterans-affairs/notification-api/actions/runs/9009041266)
[Create And Post Tag with Summary](https://github.com/department-of-veterans-affairs/notification-api/actions/runs/9009041269)

image

## Checklist

- [x] I have assigned myself to this PR
- [x] PR has an appropriate title: #9999 - What the thing does
- [x] PR has a detailed description, including links to specific documentation
- [x] I have added the [appropriate labels](https://github.com/department-of-veterans-affairs/notification-api/blob/master/.github/release.yaml) to the PR.
- [x] I did not remove any parts of the template, such as checkboxes even if they are not used
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to any documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have ensured the latest master is merged into my branch and all checks are green prior to review
- [ ] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules
- [x] The ticket is now moved into the DEV test column

### CODE REVIEW

**Review by cris-oddball — CHANGES_REQUESTED** [review]
please delete all of your non-v tags and reset whatever is stored so that we start from a fresj slate when this PR is merged.

**Review by cris-oddball — APPROVED** [review]

File: .github/scripts/prData.js [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+    const releaseBranchSha = await fetchReleaseBranchSha(github, owner, repo);
+
+    // const checkTag = await verifyNoExistingTag(github, owner, repo, releaseBranchSha)
+
+    const labels = pullRequestData.data[0].labels.map((label) => ({
+      id: label.id,
+      name: label.name,
+      description: label.description,
+      color: label.color,
+    }));
```
Comment (kalbfled) [inline .github/scripts/prData.js:84]: Test this to be certain, but I think this is a no-op that can be deleted.  The "map" return value is identical to the input.  The output, "labels", is just a copy of "const labels = pullRequestData.data[0].labels".  You could just do this . . .

```
const labels = pullRequestData.data[0].labels
```

. . . or not even have a new variable.

One reason to make a copy is that downstream code modifies the values, and you don't want to change the original values.  In this case, there's probably a more concise way to make a copy.

File: .github/scripts/prData.js [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+  let versionParts = currentVersion.split(".").map((x) => parseInt(x, 10));
+  let appliedLabel;
+
+  if (labels.some((label) => label.name === "breaking-change")) {
+    versionParts[0] += 1;
+    versionParts[1] = 0;
+    versionParts[2] = 0;
+    appliedLabel = "breaking change";
+  } else if (
+    labels.some((label) => ["hotfix", "security", "bug"].includes(label.name))
```
Comment (kalbfled) [inline .github/scripts/prData.js:48]: Please comment that this is the "patch" case.

File: .github/scripts/prData.js [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+    throw new Error("No SHA found in the response");
+  }
+}
+
+// Function to process labels to determine new version and label
+function processLabelsAndVersion(labels, currentVersion) {
+  let versionParts = currentVersion.split(".").map((x) => parseInt(x, 10));
+  let appliedLabel;
+
+  if (labels.some((label) => label.name === "breaking-change")) {
```
Comment (kalbfled) [inline .github/scripts/prData.js:42]: Please comment that this is the "major" case.

File: .github/scripts/prData.js [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+    versionParts[2] = 0;
+    appliedLabel = "breaking change";
+  } else if (
+    labels.some((label) => ["hotfix", "security", "bug"].includes(label.name))
+  ) {
+    versionParts[2] += 1;
+    appliedLabel = labels.find((label) =>
+      ["hotfix", "security", "bug"].includes(label.name),
+    ).name;
+  } else {
```
Comment (kalbfled) [inline .github/scripts/prData.js:54]: Please comment that this is the "minor" case.

File: .github/scripts/prData.js [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+    versionParts[1] += 1;
+    versionParts[2] = 0;
+    appliedLabel = labels.find((label) => label).name; // Catch-all increment
```
Comment (kalbfled) [inline .github/scripts/prData.js:57]: Isn't this just taking the first label in the list?  If that's the case, there's no need to "find".

**Review by kalbfled — CHANGES_REQUESTED** [review]

File: .github/scripts/prData.js [hunk 6]
```diff
@@ -0,0 +1,140 @@
...
+  return {
+    newVersion: versionParts.join("."),
+    appliedLabel,
+  };
+}
+
+/**
+ * Main function to handle pull request data for a GitHub repository.
+ * @param {Object} params - Parameters including GitHub client and context.
+ * @returns {Object|null} - An object containing pull request processing results or null in case of error.
```
Comment (kalbfled) [inline .github/scripts/prData.js:100]: will return a promise because it's async

File: .github/scripts/prLabelSemver.js [hunk 7]
```diff
@@ -0,0 +1,43 @@
+// prLabelSemver.js
+const { prData } = require("./prData");
+const appendSummary = require("./actionUtils");
+const fs = require("fs");
```
Comment (kalbfled) [inline .github/scripts/prLabelSemver.js:4]: Unused import.  Please check this for all files.

File: .github/scripts/prLabelSemver.js [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+
+/**
+ * Automatically labels pull requests based on semantic versioning (semver) guidelines
+ * and appends a summary to the GitHub action step.
+ *
+ * @param {object} github - The github object providing context and operations for the pull request.
+ * @param {object} context - The context object containing metadata and states for the action run.
+ * @param {object} core - The core library with utilities for logging and error handling.
+ */
+const prLabelSemver = async ({ github, context, core }) => {
```
Comment (kalbfled) [inline .github/scripts/prLabelSemver.js:14]: You're still using lambda syntax here, and the return value should be a promise.

File: .github/scripts/preTagSummary.js [hunk 9]
```diff
@@ -0,0 +1,37 @@
+// preTagSummary.js
+// This module defines a function to generate a pre-tag release summary for GitHub pull requests.
+const { prData, getReleaseVersionValue } = require("./prData");
+const appendSummary = require("./actionUtils");
+const fs = require("fs");
```
Comment (kalbfled) [inline .github/scripts/preTagSummary.js:5]: unused import

File: .github/scripts/preTagSummary.js [hunk 10]
```diff
@@ -0,0 +1,37 @@
...
+ * Asynchronously generates and appends a pre-tag release summary to the GitHub step summary file.
+ * This function retrieves the current and proposed new release versions from a pull request data,
+ * constructs a summary of the release, and appends it to the GitHub step summary for visibility
+ * in the GitHub Actions workflow.
+ *
+ * @param {object} github - The GitHub context object, providing context like repo and owner.
+ * @param {object} context - The GitHub context object with additional pull request information.
+ * @param {object} core - The GitHub core library, used for setting action failure messages.
+ */
+const preTagSummary = async ({ github, context, core }) => {
```
Comment (kalbfled) [inline .github/scripts/preTagSummary.js:17]: lambda syntax

File: .github/scripts/preTagSummary.js [hunk 11]
```diff
@@ -0,0 +1,37 @@
...
+
+/**
+ * Asynchronously generates and appends a pre-tag release summary to the GitHub step summary file.
+ * This function retrieves the current and proposed new release versions from a pull request data,
+ * constructs a summary of the release, and appends it to the GitHub step summary for visibility
+ * in the GitHub Actions workflow.
+ *
+ * @param {object} github - The GitHub context object, providing context like repo and owner.
+ * @param {object} context - The GitHub context object with additional pull request information.
+ * @param {object} core - The GitHub core library, used for setting action failure messages.
```
Comment (kalbfled) [inline .github/scripts/preTagSummary.js:15]: returns a promise

File: .github/workflows/create-and-post-tag.yml [hunk 12]
```diff
@@ -0,0 +1,24 @@
+name: Create And Post Tag with Summary
+
+on: 
+  workflow_call:
+  push:
+    branches:
+      - 1727-cd-pipeline-versioning
```
Comment (kalbfled) [inline .github/workflows/create-and-post-tag.yml:7]: Is this for testing?  Wasn't it "master" before?  Ditto elsewhere.

File: .github/scripts/prLabelSemver.js [hunk 13]
```diff
@@ -11,7 +10,9 @@ const fs = require("fs");
  * @param {object} context - The context object containing metadata and states for the action run.
  * @param {object} core - The core library with utilities for logging and error handling.
  */
```
Comment (kalbfled) [inline .github/scripts/prLabelSemver.js:12]: Return type?  You included that elsewhere.

**Review by kalbfled — CHANGES_REQUESTED** [review]

**Review by kalbfled — DISMISSED** [review]

**Review by kalbfled — DISMISSED** [review]

File: .github/scripts/actionUtils.js [hunk 14]
```diff
@@ -0,0 +1,23 @@
+// tagSummaryUtils.js
```
Comment (cris-oddball) [inline .github/scripts/actionUtils.js:1]: If this is supposed to be the file's name, it is not, Please adjust either this line to match the filename or update the filename and anywhere else it is called.

**Review by cris-oddball — CHANGES_REQUESTED** [review]

**Review by cris-oddball — APPROVED** [review]

**Review by kalbfled — APPROVED** [review]

**Pull request merged**